File: TddKatas/src/KataPotter/BookSet.cs
using System;
using System.Collections.Generic;

namespace KataPotter
{
    public class BookSet
    {
        private static readonly decimal[] DiscountTable = new[] { 0, 1, 0.95m, 0.9m, 0.8m, 0.75m };

        private readonly IList<int> _books = new List<int>();

        public BookSet()
        {
        }

        public BookSet(int book)
        {
            _books.Add(book);
        }

        public bool CanAdd(int book)
        {
            return !_books.Contains(book);
        }

        public void Add(int book)
        {
            if (!CanAdd(book))
            {
               throw new InvalidOperationException("Book " + book + " already in this set");
            }
            _books.Add(book);
        }

        public decimal Cost()
        {
            return _books.Count * 8 * GetDiscount();
        }

        private decimal GetDiscount()
        {
            return DiscountTable[_books.Count];
        }

        public decimal CostWith(int book)
        {
            var set = new BookSet();
            foreach (var b in _books)
            {
                set.Add(b);
            }
            set.Add(book);
            return set.Cost();
        }
    }
}

File: Ruby/TddKata/KataPotter/src/book_set.rb
class BookSet
  def initialize
    @books = []     
  end

  def can_add?(book)
    return !@books.include?(book)
  end

  def add(book)
    raise 'Book already in this set' unless can_add?(book)
    @books << book
  end

  def cost
    return 8 * @books.size * discount
  end

  def cost_with(book)
    new_set = BookSet.new
    @books.each { |b| new_set.add(b) }
    new_set.add(book)
    return new_set.cost
  end

  def books
    return @books
  end
  
  private

  DISCOUNTS = { 0 => 0, 1 => 1.00, 2 => 0.95, 3 => 0.90, 4 => 0.80, 5 => 0.75 }
  
  def discount
      return DISCOUNTS[@books.size]
  end
end